src/flb_chunk_trace.c

```diff
@@ -0,0 +1,685 @@
+}
+
+static struct flb_output_instance *find_calyptia_output_instance(struct flb_config *config)
+{
+    struct mk_list *head = NULL;
+    struct flb_output_instance *output = NULL;
+
+    mk_list_foreach(head, &config->outputs) {
+        output = mk_list_entry(head, struct flb_output_instance, _head);
+        if (strcmp(output->p->name, "calyptia") == 0) {
+            return output;
+        }
+    }
+    return NULL;
+}
+
+static void log_cb(void *ctx, int level, const char *file, int line,
+                   const char *str)
+{
+    if (level == CIO_LOG_ERROR) {
+        flb_error("[trace] %s", str);
+    }
+    else if (level == CIO_LOG_WARN) {
+        flb_warn("[trace] %s", str);
+    }
+    else if (level == CIO_LOG_INFO) {
+        flb_info("[trace] %s", str);
+    }
+    else if (level == CIO_LOG_DEBUG) {
+        flb_debug("[trace] %s", str);
+    }
+}
+
+static void trace_chunk_context_destroy(struct flb_chunk_trace_context *ctxt)
+{
+    if (flb_chunk_trace_has_chunks(ctxt) == FLB_TRUE) {
+        flb_chunk_trace_set_destroy(ctxt);
+        flb_input_pause_all(ctxt->flb->config);
+        return;
+    }
+    
+    // pause all inputs, then destroy the input storage.
+    flb_input_pause_all(ctxt->flb->config);
+    // waiting for all tasks to end is key to safely stopping and destroying
+    // the fluent-bit pipeline.
+    while (flb_task_running_count(ctxt->flb->config) > 0) {
+        sleep(1);
+    }
+    flb_sds_destroy(ctxt->trace_prefix);
+    flb_stop(ctxt->flb);
+    flb_destroy(ctxt->flb);
+    flb_free(ctxt);
+}
+
+void flb_chunk_trace_context_destroy(void *input)
+{
+    struct flb_input_instance *in = (struct flb_input_instance *)input;
+    pthread_mutex_lock(&in->chunk_trace_lock);
+    if (in->chunk_trace_ctxt != NULL) {
+        trace_chunk_context_destroy(in->chunk_trace_ctxt);
+        in->chunk_trace_ctxt = NULL;
+    }
+    pthread_mutex_unlock(&in->chunk_trace_lock);
+}
+
+struct flb_chunk_trace_context *flb_chunk_trace_context_new(void *trace_input,
+                                                            const char *output_name,
+                                                            const char *trace_prefix,
+                                                            void *data, struct mk_list *props)
+{
+    struct flb_input_instance *in = (struct flb_input_instance *)trace_input;
+    struct flb_config *config = in->config;
+    struct flb_input_instance *input = NULL;
+    struct flb_output_instance *output = NULL;
+    struct flb_output_instance *calyptia = NULL;
+    struct flb_chunk_trace_context *ctx = NULL;
+    struct mk_list *head = NULL;
+    struct flb_kv *prop = NULL;
+    int ret;
+
+    if (config->enable_chunk_trace == FLB_FALSE) {
+        return NULL;
+    }
+
+    pthread_mutex_lock(&in->chunk_trace_lock);
+
+    ctx = flb_calloc(1, sizeof(struct flb_chunk_trace_context));
+    if (ctx == NULL) {
+        pthread_mutex_unlock(&in->chunk_trace_lock);
+        flb_errno();
```

Comment:
flb_errno() uses errno, so it should be called right after the exception. Other libc calls might mislead the error message